litestar/_signature/models/msgspec_signature_model.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import re
+from typing import TYPE_CHECKING, Any, Optional, Union
+
+from msgspec import NODEFAULT, Meta, Struct, ValidationError, convert, defstruct
+from msgspec.structs import asdict
+from pydantic import ValidationError as PydanticValidationError
+from typing_extensions import Annotated
+
+from litestar._signature.field import SignatureField
+from litestar.params import DependencyKwarg
+from litestar.serialization import dec_hook
+from litestar.types.empty import Empty
+from litestar.utils import make_non_optional_union
+from litestar.utils.dataclass import simple_asdict
+from litestar.utils.typing import unwrap_union
+
+from .base import SignatureModel
+
+if TYPE_CHECKING:
+    from litestar.connection import ASGIConnection
+    from litestar.utils.signature import ParsedSignature
+
+    from .base import ErrorMessage
+
+__all__ = ("MsgspecSignatureModel",)
+
+MSGSPEC_CONSTRAINT_FIELDS = (
+    "gt",
+    "ge",
+    "lt",
+    "le",
+    "multiple_of",
+    "pattern",
+    "min_length",
+    "max_length",
+)
+
+ERR_RE = re.compile(r"`\$\.(.+)`$")
+
+
+class MsgspecSignatureModel(SignatureModel, Struct):
+    """Model that represents a function signature that uses a msgspec specific type or types."""
+
+    @classmethod
+    def parse_values_from_connection_kwargs(cls, connection: ASGIConnection, **kwargs: Any) -> dict[str, Any]:
+        """Extract values from the connection instance and return a dict of parsed values.
+
+        Args:
+            connection: The ASGI connection instance.
+            **kwargs: A dictionary of kwargs.
+
+        Raises:
+            ValidationException: If validation failed.
+            InternalServerException: If another exception has been raised.
+
+        Returns:
+            A dictionary of parsed values
+        """
+        messages: list[ErrorMessage] = []
+        try:
+            return convert(kwargs, cls, strict=False, dec_hook=dec_hook).to_dict()
+        except PydanticValidationError as e:
+            for exc in e.errors():
+                keys = [str(loc) for loc in exc["loc"]]
+                message = super()._build_error_message(keys=keys, exc_msg=exc["msg"], connection=connection)
+                messages.append(message)
+            raise cls._create_exception(messages=messages, connection=connection) from e
```

Comment:
This is leveraging the fact that the pydantic error is raised out of the dec hook to format the exception as we otherwise do.

After https://github.com/jcrist/msgspec/issues/456 we will hopefully be able to access the pydantic exception type from an exception chain.

We will also need to make this behavior conditional upon whether pydantic is present in environment or not.

Reply:
> After https://github.com/jcrist/msgspec/issues/456 we will hopefully be able to access the pydantic exception type from an exception chain.

This has been resolved in https://github.com/jcrist/msgspec/pull/460. If you want to try it out before the next release, you can [install msgspec from GitHub](https://jcristharif.com/msgspec/install.html#installing-from-github).

Reply:
Thank you @jcrist